File: playground/kotlin/src/main/kotlin/com/algolia/playground/Recommend.kt
package com.algolia.playground

import com.algolia.client.api.RecommendClient
import com.algolia.client.configuration.ClientOptions
import com.algolia.client.model.recommend.GetRecommendationsParams
import com.algolia.client.model.recommend.FbtModel
import com.algolia.client.model.recommend.BoughtTogetherQuery
import io.github.cdimascio.dotenv.Dotenv
import io.ktor.client.plugins.logging.*
import kotlin.system.exitProcess

suspend fun main() {

    val dotenv = Dotenv.configure().directory("../").load()

    val client = RecommendClient(
        appId = dotenv["ALGOLIA_APPLICATION_ID"],
        apiKey = dotenv["ALGOLIA_SEARCH_KEY"],
        options = ClientOptions(logLevel = LogLevel.BODY),
    )

    val searchIndex = dotenv["SEARCH_INDEX"]

    val recommendations = client.getRecommendations(
        getRecommendationsParams = GetRecommendationsParams(
            requests = listOf(
              BoughtTogetherQuery(
                    indexName = searchIndex,
                    model = FbtModel.BoughtTogether,
                    objectID = "6445156",
                    threshold = 0.0,
                )
            )
        )
    )
    println(recommendations)

    exitProcess(0)
}

File: playground/dart/lib/recommend.dart
import 'package:algolia_client_recommend/algolia_client_recommend.dart';
import 'package:dotenv/dotenv.dart';

void main() async {
  final dotenv = DotEnv()..load(['../.env']);

  // Creating an instance of the recommend client with the provided App ID and API key.
  final client = RecommendClient(
    appId: dotenv['ALGOLIA_APPLICATION_ID']!,
    apiKey: dotenv['ALGOLIA_SEARCH_KEY']!,
  );

  // Creating recommendation requests for different products.
  var requests = [
    RelatedProducts(
      model: RelatedModel.relatedProducts,
      objectID: '6445156',
      fallbackParameters: FallbackParams(
        query: 'iphone',
      ),
    ),
    RelatedProducts(
      model: RelatedModel.relatedProducts,
      objectID: '6443034',
    )
  ];

  // Fetching recommendations from Algolia using the above requests.
  final response = await client.getRecommendations(
    getRecommendationsParams: GetRecommendationsParams(requests: requests),
  );

  // Printing the recommendations.
  printRecommendations(response);

  // Close the client and dispose of all underlying resources.
  client.dispose();
}

/// Prints the search hits.
void printRecommendations(GetRecommendationsResponse response) {
  final results = response.results;

  // Loop over each result and map over the search hits,
  // converting each hit to a product.
  for (final result in results) {
    final hits = result.hits.map((e) => product(e));
    for (final (name, brand) in hits) {
      print("* $name ($brand)");
    }
  }
}

/// Converts a JSON object into a product tuple of (name, brand).
/// A data class with json deserialization can also be used.
(String, String) product(Map<String, dynamic> json) =>
    (json['name'] as String, json['brand'] as String);
